kong/runloop/handler.lua

```diff
@@ -383,37 +744,87 @@ local function register_events()
       end
 
       local ok, err = concurrency.with_coroutine_mutex(RECONFIGURE_OPTS, function()
+        -- below you are encouraged to yield for cooperative threading
+
         local rebuild_balancer = balancer_hash == nil or balancer_hash ~= current_balancer_hash
         if rebuild_balancer then
+          log(DEBUG, "stopping previously started health checkers on worker #", worker_id)
           balancer.stop_healthcheckers(CLEAR_HEALTH_STATUS_DELAY)
         end
 
-        kong.core_cache:purge()
-        kong.cache:purge()
-
         kong.default_workspace = default_ws
-        ngx.ctx.workspace = kong.default_workspace
+        ngx.ctx.workspace = default_ws
+
+        local router, err
+        if router_hash == nil or router_hash ~= current_router_hash then
```

Comment:
Is there any reason we change the rebuild order to balancers, routers, and then plugins?

Reply:
The order is:
1. stop current healtcheckers (if there are balancer changes) (manipulates global state)
2. build router (if there are router changes) (does not manipulate global state)
3. build plugins iterator (if there are plugins changes) (does not manipulate global state)

And then the non-yieding part of it:
1. flush kong cache and kong core cache
2. flush  router cache (if there were router changes)
3. switch over the router (if there were router changes)
4. switch over the plugins iterator (if there were plugin changes)
5. reinitialize balancer

The balancer is biggest culprit, we need to make it so that we can create instances of it, so that we can initiate a new one in background. Of course incremental etc. things come to play with this too. The point is that a single change set, we need to apply atomically so that all changes on proxy side are applied the same time (non-yieding).

The order is more like what we use elsewhere, like in init etc. now. Shouldn't matter too much. I felt it is cleaner this way.